### ISSUE TITLE
Collection matchers need to support nullabililty

### ISSUE THREAD

**Issue by dalewking** [kotest/kotest#2039]
I have a collection variable that is nullable type and I would like to verify that it:
- is not null
- is empty

But the problem is that beEmpty only works on non-nullable collections and will not compile with a nullable collection

The only way I have found to do it is:
```
    val actual: List<Foo>? = foo()
    actual.apply {
        shouldNotBeNull()
        shouldBeEmpty()
    }
```

beEmpty() along with all collection matchers and assertions should handle nullable lists and I should be able to just say either of these:
```
    actual.shouldBeEmpty()
    actual.should beEmpty()
```

Currently the only thing in collections that supports nullability is the `containExactlyInAnyOrder` matcher through the use of the neverNullMatcher, but that should be added to all collections matchers and assertions (and probably many other parts of the library).

**Comment by sksamuel** [comment]
What about if we added something like `mycollection.shouldBeEmptyOrNull()` ?

**Comment by sksamuel** [comment]
Yes I agree. It should work on nullables and error if it's null.

Very quick to add. In the meantime, you can do:

```kotlin
val actual: List<Foo>? = foo()
actual.shouldNotBeNull()
actual.shouldBeEmpty()
```

Because shouldNotBeNull will smart cast for you.

**Comment by dalewking** [comment]
Which is the same as what I did using an apply instead of creating a variable. My example code above showed a variable named actual just so I could emphasize the nullability of the type, but in my actual code I did not create a variable so was more like:
```
    foo().apply {
        shouldNotBeNull()
        shouldBeEmpty()
    }
```

which brings up the point that I really wish you supported a fluent interface and all methods like shouldNotBeNull returned the instance it was applied to so you could chain like this:

```
    foo()
        .shouldNotBeNull()
        .shouldBeEmpty()
```

**Comment by sksamuel** [comment]
I've updated the shouldBeEmpty and shouldNotBeEmpty matchers to support nullable references. 

eg

```kotlin
          val maybeList: List<Int>? = listOf(1)
          maybeList.shouldNotBeEmpty().shouldHaveSize(1)
```

Both those matchers, and `shouldNotBeNull` already allow chaining, eg

```kotlin
            val a: String? = "foo"
            a.shouldNotBeNull()
               .shouldHaveLength(3)
               .shouldStartWith("fo")
               .shouldEndWith("oo")
```


https://github.com/kotest/kotest/pull/2055

**Comment by dalewking** [comment]
> Both those matchers, and shouldNotBeNull already allow chaining, eg

Ah, I wasn't aware of that because that was not the case for the version of kotest we were using when I wrote those tests.

The ability to chain with shouldNotBeNull definitely makes the need for this request lower priority but still a good idea.

**Comment by sksamuel** [comment]
Yep, the PR is already up and will be merged for 4.4.1
